packages/cli/src/cli/basemaps/tileset.config.ts

```diff
@@ -0,0 +1,127 @@
+import { EpsgCode } from '@basemaps/geo';
+import * as z from 'zod';
+
+export const ImageryConfigDefaults = {
+    priority: 1000,
+    minZoom: 0,
+    maxZoom: 32,
+};
+
+const zZoom = z.number().refine((val) => val >= ImageryConfigDefaults.minZoom && val <= ImageryConfigDefaults.maxZoom, {
+    message: `must be between ${ImageryConfigDefaults.minZoom} and ${ImageryConfigDefaults.maxZoom}`,
+});
+
+const zImageryRule = z
+    .object({
+        priority: z
+            .number()
+            .refine((val) => val >= -1 && val <= 10000, { message: 'must be between -1 and 10000' })
+            .optional(),
+        minZoom: zZoom.optional(),
+        maxZoom: zZoom.optional(),
+    })
+    .refine(
+        ({ minZoom, maxZoom }) =>
+            (minZoom ?? ImageryConfigDefaults.minZoom) <= (maxZoom ?? ImageryConfigDefaults.maxZoom),
+        {
+            message: 'minZoom may no be greater than maxZoom',
+            path: ['minZoom'],
+        },
+    );
+
+const zImageryConfig = zImageryRule.extend({
+    id: z.string(),
+    name: z.string(),
+});
+
+const zBackground = z
+    .string()
+    .refine((val) => /^[0-9a-f]{8}$/.test(val), { message: 'must have 8 lower case hex characters' });
+
+const zImageryDefaultConfig = zImageryRule.extend({
+    nameContains: z.string().optional(),
+});
+
+const zProjectionConfig = z.object({
+    name: z.string(),
+    projection: z.nativeEnum(EpsgCode),
+    background: zBackground,
+    defaults: z.array(zImageryDefaultConfig).optional(),
+    imagery: z.array(zImageryConfig),
+});
+
+/**
+ * The Configuration of one Imagery set
+ */
+export type ImageryConfig = z.infer<typeof zImageryConfig>;
+
+/**
+ *  The Configuration for all the imagery in a TileSet
+ */
+export type ProjectionConfig = z.infer<typeof zProjectionConfig>;
+
+export function assertTileSetConfig(json: any): asserts json is ProjectionConfig {
+    zProjectionConfig.parse(json);
+}
+
+export interface FullImageryConfig {
+    id: string;
+    name: string;
+    priority: number;
+    minZoom: number;
+    maxZoom: number;
+}
+
+export type ImageryDefaultConfig = z.infer<typeof zImageryDefaultConfig>;
+
+const defaultFields = ['priority', 'minZoom', 'maxZoom'];
+
+export function addDefaults(defaults: ImageryDefaultConfig[], rule: ImageryConfig): FullImageryConfig {
+    const ans = { ...rule } as any;
```

Comment:
Can this be done with out the any

Reply:
I can't see how without using hardcoded if statements but if you know how I am all ears.

Reply:
I have made sure that the unit tests will pick up any deviation from the intended types